irohad/model/converters/impl/json_block_factory.cpp

```diff
@@ -69,50 +64,32 @@ namespace iroha {
-        for (auto it = signatures.Begin(); it != signatures.End(); ++it) {
-          Document signature_document;
-          auto &allocator = signature_document.GetAllocator();
-          signature_document.CopyFrom(*it, allocator);
-          auto signature = deserializeSignature(signature_document);
-          if (not signature.has_value()) {
-            log_->error("Signature parsing failure");
-            return nonstd::nullopt;
-          }
-          block.sigs.emplace_back(signature.value());
-        }
-
-        block.created_ts = document["created_ts"].GetUint64();
-        block.height = document["height"].GetUint64();
-        block.txs_number = static_cast<decltype(block.txs_number)>(
-            document["txs_number"].GetUint());
-        hexstringToArray(document["hash"].GetString(), block.hash);
-        hexstringToArray(document["prev_hash"].GetString(), block.prev_hash);
-        hexstringToArray(document["merkle_root"].GetString(),
-                         block.merkle_root);
-
-        auto& transactions = document["transactions"];
-        for (auto it = transactions.Begin(); it != transactions.End(); ++it) {
-          Document transaction_document;
-          auto &allocator = transaction_document.GetAllocator();
-          transaction_document.CopyFrom(*it, allocator);
-          auto transaction = factory_.deserialize(transaction_document);
-          if (not transaction.has_value()) {
-            log_->error("Transaction parsing failure");
-            return nonstd::nullopt;
-          }
-          block.transactions.emplace_back(transaction.value());
-        }
-
-        return block;
+        auto des = makeFieldDeserializer(document);
+        return nonstd::make_optional<model::Block>()
+            | des.Uint64(&Block::created_ts, "created_ts")
+            | des.Uint64(&Block::height, "height")
+            | des.Uint(&Block::txs_number, "txs_number")
+            | des.String(&Block::hash, "hash")
+            | des.String(&Block::prev_hash, "prev_hash")
+            | des.String(&Block::merkle_root, "merkle_root")
+            | des.Array(&Block::sigs, "signatures")
+            | des.Array(&Block::transactions, "transactions", [this](
+                                                                   auto array) {
+                 return std::accumulate(
+                     array.begin(), array.end(),
+                     nonstd::make_optional<Block::TransactionsType>(),
+                     [this](auto init, auto &x) {
+                       return init
+                           | [this, &x](auto transactions) {
+                               return factory_.deserialize(x)
+                                   | [&transactions](auto transaction) {
+                                       transactions.push_back(transaction);
+                                       return nonstd::make_optional(
+                                           transactions);
```

Comment:
Looks a bit complicated, maybe exctract the most nested lambda to some variable?

Reply:
Done.